aptos-move/block-executor/src/proptest_types/types.rs

```diff
@@ -620,22 +712,44 @@ impl<V: Into<Vec<u8>> + Arbitrary + Clone + Debug + Eq + Sync + Send> Transactio
     >(
         self,
         universe: &[K],
-        // Are writes and reads module access (same access path).
-        module_access: (bool, bool),
     ) -> MockTransaction<KeyType<K>, E> {
-        let is_module_read = |_| -> bool { module_access.1 };
-        let is_module_write = |_| -> bool { module_access.0 };
-        let is_delta = |_, _: &V| -> Option<DeltaOp> { None };
-        // Module deletion isn't allowed.
-        let allow_deletes = !(module_access.0 || module_access.1);
+        self.new_mock_write_txn(universe, true, None)
+    }
 
-        self.new_mock_write_txn(
-            universe,
-            &is_module_read,
-            &is_module_write,
-            &is_delta,
-            allow_deletes,
-        )
+    pub(crate) fn materialize_modules<
+        K: Clone + Hash + Debug + Eq + Ord,
+        E: Send + Sync + Debug + Clone + TransactionEvent,
+    >(
+        self,
+        universe: &[K],
+    ) -> MockTransaction<KeyType<K>, E> {
+        let universe_len = universe.len();
+        assert_ge!(universe_len, 3, "Universe must have size >= 3");
+
+        let mut behaviors = self
+            .new_mock_write_txn(universe, false, None)
+            .into_behaviors();
+
+        behaviors.iter_mut().for_each(|behavior| {
+            // Handle writes
+            let (key_to_convert, mut value, _) = behavior.writes.pop().unwrap();
+            let module_id = key_to_module_id(&key_to_convert, universe_len);
+
+            // Serialize a module and store it in bytes so deserialization can succeed.
+            let mut serialized_bytes = Vec::new();
+            Module::new_for_test(module_id.clone())
+                .serialize(&mut serialized_bytes)
+                .expect("Failed to serialize compiled module");
+            value.bytes = Some(serialized_bytes.into());
+
+            behavior.module_writes = vec![ModuleWrite::new(module_id, value)];
+
+            // Handle reads
+            let (key_to_convert, _) = behavior.reads.pop().unwrap();
+            behavior.module_reads = vec![key_to_module_id(&key_to_convert, universe_len)];
```

Comment:
This is the same module id as the write? can re-use module ID?

Reply:
This way we access the information that was updated by module publishing. Basically keys that are transformed as module keys are both read and written to test module publishing flow. The transformation into group keys and module keys is actually very consistent with builder pattern once we refactor it.